### PR TITLE
Feature/overrides #80

### PR DESCRIPTION
**Pull request by m-suchorski**
Not sure if it's the correct approach to this task - the fact that there are hash values declared this way doesn't look good. These values should be updated every time something in these functions changes. If there's a better way to make it work please let me know.

I also wanted to specify where is the warning coming from but I can't seem to be able to access the component's name to pass it as a second variable to the warn function - if there's a way to do that let me know.

PR regarding issue #80

### CODE REVIEW

**Comment by xolir** [comment]
I would not merge it until createDecorator will be implemented, as usage of strudel-redux will print errors because that's only way to access $teardown right now

**Comment by m-suchorski** [comment]
Somehow CircleCI tests failed, while local tests work fine, I'll have to look more into that.

**Comment by mateuszluczak** [comment]
Hmm, for me it's a sledgehammer to crack a nut.

I thought about something more simple - we don't need to do runtime check, it's more to help developers to don't do anything stupid in components.

My proposed solution would be to:
1. Specify a list of protected properties - it should include `constructor`, `$teardown`, `$off` and `$emit`
2. When `@Component` decorator is used during mixing prototypes check what properties are protected and if component tries to mix property that is protected:
   - don't mixin the property
    - warn (this should not happen for constructor property, let's make omitting it silent) that "component tried to override instance method (provide name) in component (name)"

**Comment by m-suchorski** [comment]
I haven't though of that, I'll try to do it now and I'll let you guys know if there's any problem.

**Comment by m-suchorski** [comment]
I've changed the logic, looks way simpler now.

File: examples/overrides/index.html [hunk 1]
```diff
@@ -0,0 +1,36 @@
+<body>
+  <div class="static">
```
Comment (mateuszluczak) [inline examples/overrides/index.html]: I dont think we need e2e test for that, unit test covers 100% code

File: src/util/helpers.js [hunk 2]
```diff
@@ -17,14 +19,26 @@ export const mixPrototypes = (target, source) => {
...
+
   Object.getOwnPropertyNames(inst).forEach((name) => {
     const desc = Object.getOwnPropertyDescriptor(inst, name);
     desc.writable = true;
     Object.defineProperty(targetProto, name, desc);
   });
 
   Object.getOwnPropertyNames(sourceProto).forEach((name) => {
-    if (name !== 'constructor') {
+    if (protectedMethods.includes(name)) {
```
Comment (mateuszluczak) [inline src/util/helpers.js]: You really seem to hate IE11 :P

File: src/util/helpers.js [hunk 3]
```diff
@@ -17,14 +19,26 @@ export const mixPrototypes = (target, source) => {
...
+      if (name !== 'constructor') {
+        warn(`Component tried to override instance method ${name} in component ${source.name}`);
```
Comment (mateuszluczak) [inline src/util/helpers.js]: Remove "in component X" and just pass second parameter to `warn`

File: src/util/helpers.js [hunk 4]
```diff
@@ -17,14 +19,26 @@ export const mixPrototypes = (target, source) => {
   const sourceProto = source.prototype;
   const inst = (typeof source === 'object') ? source : new source(); // eslint-disable-line new-cap
 
+  const protectedMethods = [
```
Comment (mateuszluczak) [inline src/util/helpers.js]: I would move this const outside of this function, its being interpreted every time its called which is not necessary

File: src/util/helpers.js (hunk 2 above)
Comment (m-suchorski) [reply src/util/helpers.js]: I do, actually 😄

File: examples/overrides/index.html (hunk 1 above)
Comment (m-suchorski) [reply examples/overrides/index.html]: Done.

File: src/util/helpers.js (hunk 3 above)
Comment (m-suchorski) [reply src/util/helpers.js]: Done.

File: src/util/helpers.js (hunk 2 above)
Comment (m-suchorski) [reply src/util/helpers.js]: Changed that, just want to let you know that there's still an error on IE11, it says that Object doesn't have a find method or something like that. I can fix that in another ticket.

File: src/util/helpers.js (hunk 4 above)
Comment (m-suchorski) [reply src/util/helpers.js]: Done.

File: src/util/helpers.js [hunk 5]
```diff
@@ -1,3 +1,13 @@
+import { warn } from './error';
+
+const protectedMethods = [
```
Comment (mateuszluczak) [inline src/util/helpers.js]: Can you jsdoc that?

File: src/util/helpers.js (hunk 2 above)
Comment (mateuszluczak) [reply src/util/helpers.js]: Sure!
Nice catch, I wonder what have introduced that error

Comment (mateuszluczak) [reply src/util/helpers.js]: Ok, found it https://github.com/strudeljs/strudel/blob/6fa28854a6714740bcdede12cea2268c3b26c836/src/core/init.js#L40

@xolir FYI 🔥

File: src/util/helpers.js (hunk 5 above)
Comment (m-suchorski) [reply src/util/helpers.js]: Sure thing.

**Pull request merged**